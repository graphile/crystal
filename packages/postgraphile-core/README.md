# postgraphile-core

<span class="badge-patreon"><a href="https://patreon.com/benjie" title="Support Graphile development on Patreon"><img src="https://img.shields.io/badge/sponsor-via%20Patreon-orange.svg" alt="Patreon sponsor button" /></a></span>
[![Discord chat room](https://img.shields.io/discord/489127045289476126.svg)](http://discord.gg/graphile)
[![Package on npm](https://img.shields.io/npm/v/postgraphile-core.svg?style=flat)](https://www.npmjs.com/package/postgraphile-core)
![MIT license](https://img.shields.io/npm/l/postgraphile-core.svg)
[![Follow](https://img.shields.io/badge/twitter-@GraphileHQ-blue.svg)](https://twitter.com/GraphileHQ)

This module is the compatibility between the web layer of
[PostGraphile](https://graphile.org/postgraphile/) and the GraphQL schema built
with Graphile Engine. It loads the relevant `graphile-build-pg` plugins and
augments the inflector depending on the PostGraphile options provided.

<!-- SPONSORS_BEGIN -->

## Crowd-funded open-source software

To help us develop this software sustainably under the MIT license, we ask all
individuals and businesses that use it to help support its ongoing maintenance
and development via sponsorship.

### [Click here to find out more about sponsors and sponsorship.](https://www.graphile.org/sponsor/)

And please give some love to our featured sponsors ðŸ¤©:

<table><tr>
<td align="center"><a href="http://chads.website"><img src="https://graphile.org/images/sponsors/chadf.png" width="90" height="90" alt="Chad Furman" /><br />Chad Furman</a> *</td>
<td align="center"><a href="https://storyscript.io/?utm_source=postgraphile"><img src="https://graphile.org/images/sponsors/storyscript.png" width="90" height="90" alt="Storyscript" /><br />Storyscript</a> *</td>
<td align="center"><a href="https://postlight.com/?utm_source=graphile"><img src="https://graphile.org/images/sponsors/postlight.png" width="90" height="90" alt="Postlight" /><br />Postlight</a> *</td>
</tr></table>

<em>\* Sponsors the entire Graphile suite</em>

<!-- SPONSORS_END -->

## Should you be here?

Unless you want to use the low-level API you probably want to go to the
PostGraphile (previously 'PostGraphQL') repository instead:
[https://github.com/graphile/postgraphile](https://github.com/graphile/postgraphile)

It's suitable to use this module in your own application, but please be aware
you need to bring your own security in the form of an authenticated `pgClient`
(see below).

For more information about PostGraphile and Graphile Engine please see the
documentation at [graphile.org](https://www.graphile.org/).

## `createPostGraphileSchema(pgConfig, schemas, options)`

This is the function you're most likely to use in production, it will return a
promise to a GraphQL schema. You are responsible in for implementing security by
passing a pre-authenticated `pgClient` inside the GraphQL `context` when you
resolve a GraphQL query or mutation.

Example:

```js
const schema = await createPostGraphileSchema(
  process.env.DATABASE_URL,
  ["users_schema", "posts_schema"],
  {
    dynamicJson: true,
    pgJwtSecret: process.env.JWT_SECRET,
    pgJwtTypeIdentifier: "users_schema.jwt_type",
  },
);
```

Full example:

```js
const { createPostGraphileSchema } = require("postgraphile-core");
const { graphql } = require("graphql");
const pg = require("pg");

// Create a postgres pool for efficiency
const pgPool = new pg.Pool({
  connectionString: process.env.DATABASE_URL,
});

async function runQuery(query, variables) {
  // Generate our schema using the default plugins against DATABASE_URL,
  // introspecting the two schemas specified with the options provided.
  //
  // Normally for performance you'd only do this once for your entire
  // application run, not once per query as it is here.
  const schema = await createPostGraphileSchema(
    process.env.DATABASE_URL,
    ["app_public"],
    {
      dynamicJson: true,
      pgJwtSecret: process.env.JWT_SECRET,
      pgJwtTypeIdentifier: "users_schema.jwt_type",
    },
  );

  // Fetch a postgres client from the pool
  const pgClient = await pgPool.connect();

  // Start a transaction so we can apply settings local to the transaction
  await pgClient.query("begin");

  try {
    // The following statement is equivalent to (but faster than):
    //    await pgClient.query("set local role to 'postgraphile_user'");
    //    await pgClient.query("set local jwt.claims.user_id to '27'");
    await pgClient.query(`select
      set_config('role', 'postgraphile_user', true),
      set_config('jwt.claims.user_id', '27', true)
    `);
    return await graphql(
      schema,
      query,
      null,
      /* CONTEXT > */ {
        pgClient: pgClient,
      } /* < CONTEXT */,
      variables,
    );
  } finally {
    // commit the transaction (or rollback if there was an error) to clear the local settings
    await pgClient.query("commit");

    // Release the pgClient back to the pool.
    await pgClient.release();
  }
}

// Normally you'd execute a query in response to an HTTP request or similar
runQuery(
  // This query obviously depends on your database schema
  "query MyQuery { allPosts { nodes { id, title, author: userByAuthorId { username } } } }",
)
  .then((result) => {
    console.dir(result);
    pgPool.end();
  })
  .catch((e) => {
    console.error(e);
    process.exit(1);
  });
```

To see how this works in a real application, check out
[`withPostGraphileContext` in PostGraphile](https://github.com/graphile/postgraphile/blob/master/src/postgraphile/withPostGraphileContext.ts)

## `watchPostGraphileSchema(pgConfig, schemas, options, onNewSchema)`

This function is useful in development; it returns a promise that resolves to a
`release` function that you can call to stop watching. The `onNewSchema`
callback will be called every time a new schema is generated, and it is
guaranteed to be called before the returned promise resolves. Other than the
additional `onNewSchema` option, the options are identical to that of
`createPostGraphileSchema` above.
