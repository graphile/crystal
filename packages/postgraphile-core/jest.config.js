const rootConfig = require("../../jest.config.base")(__dirname);

module.exports = {
  ...rootConfig,
  testMatch: [...rootConfig.testMatch, "**/*.test.graphql"],
  transform: {
    ...rootConfig.transform,
    "^.+\\.graphql$": "<rootDir>/__tests__/transform-graphql.ts",
  },
  moduleFileExtensions: [...rootConfig.moduleFileExtensions, "graphql"],
};
