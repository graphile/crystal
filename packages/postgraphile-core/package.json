{
  "name": "postgraphile-core",
  "version": "4.7.0",
  "description": "",
  "main": "dist/index.js",
  "scripts": {
    "test": "jest -i",
    "prepack": "tsc -b"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/graphile/graphile-build.git"
  },
  "author": "Benjie Gillam <code@benjiegillam.com>",
  "homepage": "https://github.com/graphile/graphile-build/tree/master/packages/postgraphile-core",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/graphile/graphile-build/issues"
  },
  "dependencies": {
    "@types/jsonwebtoken": "^8.5.0",
    "graphile-build": "4.7.0",
    "graphile-build-pg": "4.7.0",
    "graphql-parse-resolve-info": "4.7.0",
    "tslib": "^2.0.0"
  },
  "devDependencies": {
    "@graphile/subscriptions-lds": "4.7.0",
    "@types/debug": "^4.1.5",
    "@types/jest": "^25.2.3",
    "@types/json5": "0.0.30",
    "@types/mock-fs": "^4.10.0",
    "@types/node": "^14.0.1",
    "@types/pg": "^7.14.3",
    "debug": "^4.2.0",
    "graphile-utils": "4.7.0",
    "graphql": "^14.6.0",
    "jest": "^26.0.1",
    "json5": "^2.1.3",
    "jsonwebtoken": "^8.5.1",
    "mock-fs": "^4.12.0",
    "pg": "^8.2.1",
    "pg-connection-string": "^2.2.3",
    "ts-node": "^8.10.1",
    "typescript": "^3.9.2"
  },
  "peerDependencies": {
    "graphql": "^14.6.0",
    "pg": ">=6.1.0 <9"
  },
  "files": [
    "dist",
    "index.js"
  ],
  "engines": {
    "node": ">=12.16"
  }
}
