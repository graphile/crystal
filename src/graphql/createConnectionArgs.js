import { memoize, fromPairs, includes, snakeCase, toUpper } from 'lodash'
import { GraphQLEnumType, GraphQLInt, GraphQLBoolean } from 'graphql'
import { CursorType } from './types.js'
import getType from './getType.js'

const createConnectionArgs = (table, ignoreColumnConditions = []) => ({
  // The column specified by `orderBy` means more than just the order to
  // return items in. This column is also the column we will use for
  // cursors.
  orderBy: {
    type: createTableOrderingEnum(table),
    description:
      'The order the resulting items should be returned in. This argument ' +
      'is also important as it is used in creating pagination cursors. This ' +
      'value’s default is the primary key for the object.',
    defaultValue: (() => {
      const column = table.getPrimaryKeys()[0]
      if (column) return column.name
      return null
    })(),
  },
  first: {
    type: GraphQLInt,
    description:
      'The top `n` items in the set to be returned. Can’t be used ' +
      'with `last`.',
  },
  last: {
    type: GraphQLInt,
    description:
      'The bottom `n` items in the set to be returned. Can’t be used ' +
      'with `first`.',
  },
  before: {
    type: CursorType,
    description: 'Constrains the set to nodes *before* this cursor in the specified ordering.',
  },
  after: {
    type: CursorType,
    description: 'Constrains the set to nodes *after* this cursor in the specified ordering.',
  },
  offset: {
    type: GraphQLInt,
    description: 'An integer offset representing how many items to skip in the set.',
  },
  descending: {
    type: GraphQLBoolean,
    description:
      'If `true` the nodes will be in descending order, if `false` the ' +
      'items will be in ascending order. `false` by default.',
    defaultValue: false,
  },
  ...fromPairs(
    table
    .getColumns()
    // If `ignoreColumnConditions` is set to true, all column conditions will
    // be disabled. If `ignoreColumnConditions` is an array, only certain
    // conditions will be ignored.
    .filter(column => (ignoreColumnConditions === true ? false : !includes(ignoreColumnConditions, column)))
    .map(column => [column.getFieldName(), {
      type: getType(column.type),
      description:
        'Filters the resulting set with an equality test on the ' +
        `${column.getMarkdownFieldName()} field.`,
      // TODO: Deprecate this…
      // deprecationReason:
      //   'Simple equality testing is insufficient for the nodes field and just ' +
      //   'adds noise. Instead use procedures for custom set filtering.',
    }])
  ),
})

export default createConnectionArgs

/**
 * Creates an ordering enum which simply contains all of a `Table`s columns.
 *
 * @param {Table} table
 * @returns {GraphQLEnumType}
 */
// TODO: Some way to eliminate some columns from ordering enum?
export const createTableOrderingEnum = memoize(table =>
  new GraphQLEnumType({
    name: `${table.getTypeName()}Ordering`,
    description: `Properties with which ${table.getMarkdownTypeName()} can be ordered.`,

    values: fromPairs(
      table
      .getColumns()
      .map(column => [toUpper(snakeCase(column.getFieldName())), {
        value: column.name,
        description: column.description,
      }])
    ),
  }))
