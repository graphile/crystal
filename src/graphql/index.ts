import createGraphQLSchema from './schema/createGQLSchema'

export {
  createGraphQLSchema,
}
