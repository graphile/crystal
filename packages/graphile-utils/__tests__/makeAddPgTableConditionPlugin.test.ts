import pg, { Pool } from "pg";
import { graphql } from "graphql";
import { createPostGraphileSchema } from "postgraphile-core";
import {
  makeExtendSchemaPlugin,
  gql,
  embed,
  makeAddPgTableConditionPlugin,
} from "../dist";
import assert from "assert";

const clean = (data: any): any => {
  if (Array.isArray(data)) {
    return data.map(clean);
  } else if (data && typeof data === "object") {
    return Object.keys(data).reduce((memo, key) => {
      const value = data[key];
      if (key === "id" && typeof value === "number") {
        memo[key] = "[id]";
      } else if (key === "nodeId" && typeof value === "string") {
        memo[key] = "[nodeId]";
      } else {
        memo[key] = clean(value);
      }
      return memo;
    }, {});
  } else {
    return data;
  }
};

let pgPool: Pool;

beforeAll(() => {
  pgPool = new pg.Pool({
    connectionString: process.env.TEST_DATABASE_URL,
  });
});

afterAll(() => {
  if (pgPool) {
    pgPool.end();
  }
});

const PetsCountPlugin = makeAddPgTableConditionPlugin(
  "graphile_utils",
  "users",
  "petCountAtLeast",
  (build) => ({
    description: "Filters users to those that have at least this many pets",
    type: build.graphql.GraphQLInt,
  }),
  (value, helpers, build) => {
    expect(build.graphql).toBeTruthy();
    const { sqlTableAlias, sql } = helpers;
    return sql`(select count(*) from graphile_utils.pets where pets.user_id = ${sqlTableAlias}.id) >= ${sql.value(
      value as any,
    )}`;
  },
);

it("allows adding a condition to a Relay connection", async () => {
  const schema = await createPostGraphileSchema(pgPool, ["graphile_utils"], {
    disableDefaultMutations: true,
    appendPlugins: [PetsCountPlugin],
  });
  expect(schema).toMatchSnapshot();
  const pgClient = await pgPool.connect();
  try {
    const { data, errors } = await graphql(
      schema,
      `
        query {
          allUsers(condition: { petCountAtLeast: 3 }) {
            nodes {
              nodeId
              id
              name
              email
              bio
            }
          }
        }
      `,
      null,
      { pgClient },
      {},
    );
    expect(errors).toBeFalsy();
    expect(data).toBeTruthy();
    assert(data);
    expect(data.allUsers).toBeTruthy();
    expect(data.allUsers.nodes).toBeTruthy();
    expect(data.allUsers.nodes.length).toEqual(1);
    expect(data.allUsers.nodes[0].email).toEqual("caroline@example.com");
  } finally {
    await pgClient.release();
  }
});

it("allows adding a condition to a simple collection", async () => {
  const schema = await createPostGraphileSchema(pgPool, ["graphile_utils"], {
    disableDefaultMutations: true,
    simpleCollections: "both",
    appendPlugins: [PetsCountPlugin],
  });
  expect(schema).toMatchSnapshot();
  const pgClient = await pgPool.connect();
  try {
    const { data, errors } = await graphql(
      schema,
      `
        query {
          allUsersList(condition: { petCountAtLeast: 3 }) {
            nodeId
            id
            name
            email
            bio
          }
        }
      `,
      null,
      { pgClient },
      {},
    );
    expect(errors).toBeFalsy();
    expect(data).toBeTruthy();
    assert(data);
    expect(data.allUsersList).toBeTruthy();
    expect(data.allUsersList.length).toEqual(1);
    expect(data.allUsersList[0].email).toEqual("caroline@example.com");
  } finally {
    await pgClient.release();
  }
});
