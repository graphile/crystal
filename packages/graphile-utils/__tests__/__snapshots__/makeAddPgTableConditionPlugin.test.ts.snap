// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows adding a condition to a Relay connection 1`] = `
type Complex {
  numberInt: Int
  stringText: String
}

"""An input for mutations affecting \`Complex\`"""
input ComplexInput {
  numberInt: Int
  stringText: String
}

"""A location in a connection that can be used for resuming pagination."""
scalar Cursor

"""
A point in time as described by the [ISO
8601](https://en.wikipedia.org/wiki/ISO_8601) standard. May or may not include a timezone.
"""
scalar Datetime

"""An object with a globally unique \`ID\`."""
interface Node {
  """
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  """
  nodeId: ID!
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: Cursor

  """When paginating forwards, the cursor to continue."""
  endCursor: Cursor
}

type Pet implements Node {
  """
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  """
  nodeId: ID!
  id: Int!
  userId: Int!
  name: String!
  type: String!
}

"""
A condition to be used against \`Pet\` object types. All fields are tested for equality and combined with a logical ‘and.’
"""
input PetCondition {
  """Checks for equality with the object’s \`id\` field."""
  id: Int

  """Checks for equality with the object’s \`userId\` field."""
  userId: Int

  """Checks for equality with the object’s \`name\` field."""
  name: String

  """Checks for equality with the object’s \`type\` field."""
  type: String
}

"""A connection to a list of \`Pet\` values."""
type PetsConnection {
  """A list of \`Pet\` objects."""
  nodes: [Pet]!

  """
  A list of edges which contains the \`Pet\` and cursor to aid in pagination.
  """
  edges: [PetsEdge!]!

  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """The count of *all* \`Pet\` you could get from the connection."""
  totalCount: Int!
}

"""A \`Pet\` edge in the connection."""
type PetsEdge {
  """A cursor for use in pagination."""
  cursor: Cursor

  """The \`Pet\` at the end of the edge."""
  node: Pet
}

"""Methods to use when ordering \`Pet\`."""
enum PetsOrderBy {
  NATURAL
  ID_ASC
  ID_DESC
  USER_ID_ASC
  USER_ID_DESC
  NAME_ASC
  NAME_DESC
  TYPE_ASC
  TYPE_DESC
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
}

"""The root query type which gives access points into the data universe."""
type Query implements Node {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!

  """
  The root query type must be a \`Node\` to work well with Relay 1 mutations. This just resolves to \`query\`.
  """
  nodeId: ID!

  """Fetches an object given its globally unique \`ID\`."""
  node(
    """The globally unique \`ID\`."""
    nodeId: ID!
  ): Node

  """Reads and enables pagination through a set of \`Pet\`."""
  allPets(
    """Only read the first \`n\` values of the set."""
    first: Int

    """Only read the last \`n\` values of the set."""
    last: Int

    """
    Skip the first \`n\` values from our \`after\` cursor, an alternative to cursor
    based pagination. May not be used with \`last\`.
    """
    offset: Int

    """Read all values in the set before (above) this cursor."""
    before: Cursor

    """Read all values in the set after (below) this cursor."""
    after: Cursor

    """The method to use when ordering \`Pet\`."""
    orderBy: [PetsOrderBy!] = [PRIMARY_KEY_ASC]

    """
    A condition to be used in determining which values should be returned by the collection.
    """
    condition: PetCondition
  ): PetsConnection

  """Reads and enables pagination through a set of \`User\`."""
  allUsers(
    """Only read the first \`n\` values of the set."""
    first: Int

    """Only read the last \`n\` values of the set."""
    last: Int

    """
    Skip the first \`n\` values from our \`after\` cursor, an alternative to cursor
    based pagination. May not be used with \`last\`.
    """
    offset: Int

    """Read all values in the set before (above) this cursor."""
    before: Cursor

    """Read all values in the set after (below) this cursor."""
    after: Cursor

    """The method to use when ordering \`User\`."""
    orderBy: [UsersOrderBy!] = [PRIMARY_KEY_ASC]

    """
    A condition to be used in determining which values should be returned by the collection.
    """
    condition: UserCondition
  ): UsersConnection
  petById(id: Int!): Pet
  userById(id: Int!): User

  """Reads a single \`Pet\` using its globally unique \`ID\`."""
  pet(
    """The globally unique \`ID\` to be used in selecting a single \`Pet\`."""
    nodeId: ID!
  ): Pet

  """Reads a single \`User\` using its globally unique \`ID\`."""
  user(
    """The globally unique \`ID\` to be used in selecting a single \`User\`."""
    nodeId: ID!
  ): User
}

type User implements Node {
  """
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  """
  nodeId: ID!
  id: Int!
  name: String!
  email: String!
  bio: String
  renamedComplexColumn: [Complex]
  createdAt: Datetime!
}

"""
A condition to be used against \`User\` object types. All fields are tested for equality and combined with a logical ‘and.’
"""
input UserCondition {
  """Checks for equality with the object’s \`id\` field."""
  id: Int

  """Checks for equality with the object’s \`name\` field."""
  name: String

  """Checks for equality with the object’s \`email\` field."""
  email: String

  """Checks for equality with the object’s \`bio\` field."""
  bio: String

  """Checks for equality with the object’s \`renamedComplexColumn\` field."""
  renamedComplexColumn: [ComplexInput]

  """Checks for equality with the object’s \`createdAt\` field."""
  createdAt: Datetime

  """Filters users to those that have at least this many pets"""
  petCountAtLeast: Int
}

"""A connection to a list of \`User\` values."""
type UsersConnection {
  """A list of \`User\` objects."""
  nodes: [User]!

  """
  A list of edges which contains the \`User\` and cursor to aid in pagination.
  """
  edges: [UsersEdge!]!

  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """The count of *all* \`User\` you could get from the connection."""
  totalCount: Int!
}

"""A \`User\` edge in the connection."""
type UsersEdge {
  """A cursor for use in pagination."""
  cursor: Cursor

  """The \`User\` at the end of the edge."""
  node: User
}

"""Methods to use when ordering \`User\`."""
enum UsersOrderBy {
  NATURAL
  ID_ASC
  ID_DESC
  NAME_ASC
  NAME_DESC
  EMAIL_ASC
  EMAIL_DESC
  BIO_ASC
  BIO_DESC
  RENAMED_COMPLEX_COLUMN_ASC
  RENAMED_COMPLEX_COLUMN_DESC
  CREATED_AT_ASC
  CREATED_AT_DESC
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
}

`;

exports[`allows adding a condition to a simple collection 1`] = `
type Complex {
  numberInt: Int
  stringText: String
}

"""An input for mutations affecting \`Complex\`"""
input ComplexInput {
  numberInt: Int
  stringText: String
}

"""A location in a connection that can be used for resuming pagination."""
scalar Cursor

"""
A point in time as described by the [ISO
8601](https://en.wikipedia.org/wiki/ISO_8601) standard. May or may not include a timezone.
"""
scalar Datetime

"""An object with a globally unique \`ID\`."""
interface Node {
  """
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  """
  nodeId: ID!
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: Cursor

  """When paginating forwards, the cursor to continue."""
  endCursor: Cursor
}

type Pet implements Node {
  """
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  """
  nodeId: ID!
  id: Int!
  userId: Int!
  name: String!
  type: String!
}

"""
A condition to be used against \`Pet\` object types. All fields are tested for equality and combined with a logical ‘and.’
"""
input PetCondition {
  """Checks for equality with the object’s \`id\` field."""
  id: Int

  """Checks for equality with the object’s \`userId\` field."""
  userId: Int

  """Checks for equality with the object’s \`name\` field."""
  name: String

  """Checks for equality with the object’s \`type\` field."""
  type: String
}

"""A connection to a list of \`Pet\` values."""
type PetsConnection {
  """A list of \`Pet\` objects."""
  nodes: [Pet]!

  """
  A list of edges which contains the \`Pet\` and cursor to aid in pagination.
  """
  edges: [PetsEdge!]!

  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """The count of *all* \`Pet\` you could get from the connection."""
  totalCount: Int!
}

"""A \`Pet\` edge in the connection."""
type PetsEdge {
  """A cursor for use in pagination."""
  cursor: Cursor

  """The \`Pet\` at the end of the edge."""
  node: Pet
}

"""Methods to use when ordering \`Pet\`."""
enum PetsOrderBy {
  NATURAL
  ID_ASC
  ID_DESC
  USER_ID_ASC
  USER_ID_DESC
  NAME_ASC
  NAME_DESC
  TYPE_ASC
  TYPE_DESC
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
}

"""The root query type which gives access points into the data universe."""
type Query implements Node {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!

  """
  The root query type must be a \`Node\` to work well with Relay 1 mutations. This just resolves to \`query\`.
  """
  nodeId: ID!

  """Fetches an object given its globally unique \`ID\`."""
  node(
    """The globally unique \`ID\`."""
    nodeId: ID!
  ): Node

  """Reads and enables pagination through a set of \`Pet\`."""
  allPets(
    """Only read the first \`n\` values of the set."""
    first: Int

    """Only read the last \`n\` values of the set."""
    last: Int

    """
    Skip the first \`n\` values from our \`after\` cursor, an alternative to cursor
    based pagination. May not be used with \`last\`.
    """
    offset: Int

    """Read all values in the set before (above) this cursor."""
    before: Cursor

    """Read all values in the set after (below) this cursor."""
    after: Cursor

    """The method to use when ordering \`Pet\`."""
    orderBy: [PetsOrderBy!] = [PRIMARY_KEY_ASC]

    """
    A condition to be used in determining which values should be returned by the collection.
    """
    condition: PetCondition
  ): PetsConnection

  """Reads a set of \`Pet\`."""
  allPetsList(
    """Only read the first \`n\` values of the set."""
    first: Int

    """Skip the first \`n\` values."""
    offset: Int

    """The method to use when ordering \`Pet\`."""
    orderBy: [PetsOrderBy!]

    """
    A condition to be used in determining which values should be returned by the collection.
    """
    condition: PetCondition
  ): [Pet!]

  """Reads and enables pagination through a set of \`User\`."""
  allUsers(
    """Only read the first \`n\` values of the set."""
    first: Int

    """Only read the last \`n\` values of the set."""
    last: Int

    """
    Skip the first \`n\` values from our \`after\` cursor, an alternative to cursor
    based pagination. May not be used with \`last\`.
    """
    offset: Int

    """Read all values in the set before (above) this cursor."""
    before: Cursor

    """Read all values in the set after (below) this cursor."""
    after: Cursor

    """The method to use when ordering \`User\`."""
    orderBy: [UsersOrderBy!] = [PRIMARY_KEY_ASC]

    """
    A condition to be used in determining which values should be returned by the collection.
    """
    condition: UserCondition
  ): UsersConnection

  """Reads a set of \`User\`."""
  allUsersList(
    """Only read the first \`n\` values of the set."""
    first: Int

    """Skip the first \`n\` values."""
    offset: Int

    """The method to use when ordering \`User\`."""
    orderBy: [UsersOrderBy!]

    """
    A condition to be used in determining which values should be returned by the collection.
    """
    condition: UserCondition
  ): [User!]
  petById(id: Int!): Pet
  userById(id: Int!): User

  """Reads a single \`Pet\` using its globally unique \`ID\`."""
  pet(
    """The globally unique \`ID\` to be used in selecting a single \`Pet\`."""
    nodeId: ID!
  ): Pet

  """Reads a single \`User\` using its globally unique \`ID\`."""
  user(
    """The globally unique \`ID\` to be used in selecting a single \`User\`."""
    nodeId: ID!
  ): User
}

type User implements Node {
  """
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  """
  nodeId: ID!
  id: Int!
  name: String!
  email: String!
  bio: String
  renamedComplexColumn: [Complex]
  createdAt: Datetime!
}

"""
A condition to be used against \`User\` object types. All fields are tested for equality and combined with a logical ‘and.’
"""
input UserCondition {
  """Checks for equality with the object’s \`id\` field."""
  id: Int

  """Checks for equality with the object’s \`name\` field."""
  name: String

  """Checks for equality with the object’s \`email\` field."""
  email: String

  """Checks for equality with the object’s \`bio\` field."""
  bio: String

  """Checks for equality with the object’s \`renamedComplexColumn\` field."""
  renamedComplexColumn: [ComplexInput]

  """Checks for equality with the object’s \`createdAt\` field."""
  createdAt: Datetime

  """Filters users to those that have at least this many pets"""
  petCountAtLeast: Int
}

"""A connection to a list of \`User\` values."""
type UsersConnection {
  """A list of \`User\` objects."""
  nodes: [User]!

  """
  A list of edges which contains the \`User\` and cursor to aid in pagination.
  """
  edges: [UsersEdge!]!

  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """The count of *all* \`User\` you could get from the connection."""
  totalCount: Int!
}

"""A \`User\` edge in the connection."""
type UsersEdge {
  """A cursor for use in pagination."""
  cursor: Cursor

  """The \`User\` at the end of the edge."""
  node: User
}

"""Methods to use when ordering \`User\`."""
enum UsersOrderBy {
  NATURAL
  ID_ASC
  ID_DESC
  NAME_ASC
  NAME_DESC
  EMAIL_ASC
  EMAIL_DESC
  BIO_ASC
  BIO_DESC
  RENAMED_COMPLEX_COLUMN_ASC
  RENAMED_COMPLEX_COLUMN_DESC
  CREATED_AT_ASC
  CREATED_AT_DESC
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
}

`;
